### ISSUE TITLE
A planned restored suffix whose final prime segment is exactly PRIME_MIN_T (16) rows is not bitwise

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#427]
Refs #420 (qwen35 calibrated prefill activation lane, where this was found). **Not fixed there** —
that PR only documents it. Filing for the next lane.

## What

Under the render-stable boundary protocol (#248/#256/#257), a restored suffix prime is expected to
be bitwise equal to the equivalent one-call prime when the head is grid-aligned
(`head % Engine::gdn_chunk_size() == 0`). It is, at every tail size measured, EXCEPT a final prime
segment of exactly 16 rows (`PRIME_MIN_T`).

It affects the served q38 NVFP4-Q5K mint and a calibrated A4 artifact identically, so it is not
about the activation program.

## Reproducer

`qwen-a4-continuation-gate` (on `lane/qwen-fp4-activation-mint-20260909`) primes the same 9,296
tokens with no server, no prefix cache and no drafter, once in a single call and once as
head+suffix, and compares the final logits:

```
qwen-a4-continuation-gate <model.gguf> <prompt.txt> 9296 16 48 80 112 144 176 208
```

Heads are 9280, 9248, 9216, 9184, 9152, 9120, 9088 — all multiples of 32.

| split | served q38 mint | calibrated A4 artifact |
|---|---|---|
| 9280 + **16** | **DIFFERS** | **DIFFERS** |
| 9248 + 48 | ok | ok |
| 9216 + 80 | ok | ok |
| 9184 + 112 | ok | ok |
| 9152 + 144 | ok | ok |
| 9120 + 176 | ok | ok |
| 9088 + 208 | ok | ok |

## Control

`run-spec --prime-walker-check` PASSes on both artifacts on a 9,067-row real prompt (6 yields,
planned capture, resumed logits bitwise vs the unyielded trunk, boundary state digest matching),
so the protocol itself is sound and this is specific to the 16-row tail shape.

## Why it matters

It is reachable from ordinary serving, not just from a synthetic split. A 48-token restored suffix
primes as 32 + 16, so the final call lands on exactly this shape. In the lane's four-turn greedy
restore chain that produced a divergent turn on one transcript, while a chain whose suffix happened
to be 67 tokens (35 final rows) matched 4/4.

`check_mtp_prime_walker` already refuses a plan whose tail is BELOW `PRIME_MIN_T`; a tail of
exactly `PRIME_MIN_T` is accepted and is not bitwise.

## Suggested shape of a fix

Either make the 16-row segment bitwise, or have the resume splitter avoid leaving a tail of exactly
`PRIME_MIN_T` (fold it into the previous segment), or refuse the plan the way a shorter tail is
refused. Whichever it is, it wants a gate arm at tail == PRIME_MIN_T.

Receipts: darklanes `research/qwen-fp4-activation-mint-20260909/RESTORE-FENCE.md` and
`gates/walker-and-aligned.log`.

**Comment by avifenesh** [comment]
Reproduced on today's `main` (`fa5185d90` merged into `lane/spill-b-20260919`, tip `c5384d53c`) and classified. Receipts: `research/spill-b-20260919/DAY21.md` section 2, cells `rtx5090-day21/{cont-gate,cont-bisect,cont-bisect2}/` and `pro-single-day21/cont-gate/` (all `executed-not-qualified`, collector-locked, card alone before and after). No code changed.

**The two capture-site fixes (#379 in the hit gate, #602 in #606) did not change the outcome.** They move where a seed lands; they do not touch the prime program of a 16-row call. Reproducer: `qwen-a4-continuation-gate` (on `main` since #424), the served `Qwen3.8-27B-NVFP4-Q5K-mtp.gguf`, prompt `docs/SERVING.md` truncated by the gate, heads all multiples of 32, `MEMRA_PRIME_ROW_RECEIPT=1`. Local RTX 5090 Laptop GPU; the RTX PRO 6000 Blackwell printed byte-identical digests on every split it ran (16, 48, 80 at 9296; 17 at 9297; 32 at 9312):

| total | split | one-call `logits_sha` | split `logits_sha` | verdict |
| ---: | --- | --- | --- | --- |
| 9296 | 9280 + **16** | `14ab5f8b365dbd71` | `35bd15f063bfd5ba` | **DIFFERS** |
| 9296 | 9248 + 48 | `14ab5f8b365dbd71` | `14ab5f8b365dbd71` | ok |
| 9296 | 9216 + 80, 9184 + 112, 9152 + 144, 9120 + 176, 9088 + 208 | `14ab5f8b365dbd71` | `14ab5f8b365dbd71` | ok |
| 9297 | 9280 + **17** | `e641952cac19e526` | `e641952cac19e526` | ok |
| 9311 | 9280 + **31** | `b61719f294866b05` | `b61719f294866b05` | ok |
| 9312 | 9280 + **32** | `736a88d7c7448dcb` | `736a88d7c7448dcb` | ok |

Same head, tails 16, 17, 31, 32: only 16 differs. The `[prime-row]` receipts show the 48-row suffix reproducing the one-call row exactly (`top1=318 margin=1.375317e0`) and the 16-row suffix keeping the argmax while moving the margin (`margin=1.369547e0`), which is why this only flips a near-tie transcript.

**Bisect on the same binary, existing rollback seams, tails 16 and 48 at 9296.** The 16-row split still DIFFERS under `MEMRA_GDN_CHUNKED=0`, `MEMRA_F16OUT=0` (inert here, digests unchanged), `MEMRA_NOFA=1` and `MEMRA_MMQ_W4A8=0`; the 48-row split is `ok` in every arm. Read from the tree and cleared as well: `GEMM_M_THRESHOLD` is 16 with the test `m >= 16`, so a 16-row call takes the prefill GEMM like a 48-row call; the W4A8 MMQ tile is fixed at compile time; the f32 row GEMV that refuses `m > 16` is behind `MEMRA_F32_GEMV_KERNEL=1`, default off.

**Classification (the chunk-width arms).** With `MEMRA_PRIME_CHUNK=32` the ONE-CALL prime over 9296 tokens ends in a 16-row chunk of its own (9296 = 290 x 32 + 16; the fold rule folds only tails strictly below PRIME_MIN_T) and digests to **`35bd15f063bfd5ba`**, exactly the digest the 16-row suffix call produced under the default schedule, and both splits read `ok` against it. With `=16` every call in both arms is a 16-row chunk and both splits read `ok`. So the restore protocol is exact: the same rows computed as a 16-row chunk give the same bits whether that chunk is the tail of a cold prime or a restored suffix, and they differ from the same rows computed inside a wider chunk (the default schedule ends 9296 with an 80-row chunk). **The seam is that a prime chunk of exactly PRIME_MIN_T rows is its own numeric program in the per-chunk layer walk**, and every cold prime whose schedule ends in a 16-row chunk carries it too. (The `=48` arm reads DIFFERS on both splits, which is the GDN grid law, not this edge: an explicit `MEMRA_PRIME_CHUNK` bypasses `align_prime_ranges_to_gdn` and puts the one-call's boundary at 9264, off the 32-grid; reported as measured, not read as evidence.)

Which kernel changes program at width 16 is not named today: the arms clear the GDN scan, attention, f16out, the W4A8 class, the MMQ tile, the GEMM threshold and the f32 GEMV door. What remains keyed on the call width are the dense matmuls that reach cuBLASLt (shape-keyed heuristics) and the GDN prep/conv kernels' width tiling. The next cell is a per-layer, per-tensor digest of a 16-row chunk against a 17-row chunk on the same rows.

**Not fixed.** Both fix shapes change bytes outside the restore path and need their own battery, so neither fits the one-program-with-a-bit-identity-gate rule inside today's budget: (1) engine side, make a 16-row chunk take the wide-chunk program once the per-layer digest names the kernel (changes every cold prime with t = 16 mod chunk; needs kernel-check, run-gen argmax, run-spec K=1..8, chunkinv at width 16, and this gate's "known" arm turned into a verdict); (2) schedule side, never emit a 16-row segment: fold rule `<= PRIME_MIN_T`, the worker's three capture sites leave more than PRIME_MIN_T prompt tokens behind a seed (today's rule leaves at least 16, which is exactly this shape: P = 9296 publishes 9280), `DflashRestoreSuffix::new` and `check_mtp_prime_walker` refuse a suffix of exactly 16 as they refuse below it, plus the twin, restore and #379 hit gates green on both cards. Shape 2 is the smaller numeric change and closes the served surface; shape 1 is what makes the cold prime chunk-invariant at width 16; they are independent. Details in DAY21.md 2.4. Not closing.

**Comment by avifenesh** [comment]
Named and fixed on `lane/spill-b-20260919` (day 22, tip `99af46bf2`; fix commit `89c693be0`). Record: `research/spill-b-20260919/DAY22.md` (pre-registration in 1.1 to 1.4, the verbatim cells in 1.3 and 1.5, the target card in 2) and `research/spill-b-20260919/PRIME-MIN-T-DECISION.md`. Every cell `executed-not-qualified`; nothing merged; the lead integrates. Not closing.

**The operation.** `Engine::matmul` (and `matmul_pre`, `crates/memra-engine/src/lib.rs`) for a quantized projection with `out_f < GEMM_MIN_OUT_F (128)`: it skips the `m >= 16` GEMM/MMQ arms and at m = 16 lands in the batched weight-resident MMVQ tier `(2..=16).contains(&m)` (`qmatvec_mmvq_batched`, the 32-thread warp reduce), while at m >= 17 it takes the `grid.y = m` dp4a kernel (the 128-thread two-level reduce). On the served mint those are the GDN `ssm_beta`/`ssm_alpha` projections (NVFP4 `[5120, 48]`) of all 48 GDN layers. Per-operation walk (`qwen-a4-width-walk`, each prime-path projection at 17 rows versus 16 and 48, shared rows bitwise), pre-fix tree:

```text
WIDTH WALK width 16 vs 17: 96 of 497 tensors differ; tensor names: {"ssm_alpha", "ssm_beta"}; sites: ["00/ssm_beta", "00/ssm_alpha", ..., "62/ssm_beta", "62/ssm_alpha"]
WIDTH WALK width 48 vs 17: 0 of 497 tensors differ; tensor names: {}; sites: []
layer 00 ssm_beta  qtype=NVFP4 in_f=5120 out_f=48 width 16 vs 17: rows_differ=16/16 maxabs=1.490e-7 ref_sha=e40f0aeaaec76762 sha=0a984de6e2ed5fd0 DIFFERS
```

With the tier off (`MEMRA_NO_BATCHED=1`, the documented A/B reference) the 16-row split read `ok` at the one-call digest, so the tier was the only width-keyed operation in the walk.

**The fix (one dispatch change, no new program, no flag).** An RAII scope `Engine::prefill_rows_scope` (same guard type as `verify_exact`) armed at the top of `HybridModel::prime_layers` (`hybrid_forward.rs`, the per-chunk layer walk), and one admission clause on the tier in `matmul` and `matmul_pre`: `batched_tier_admits() = verify_exact_on() || !prefill_rows_on()`. A 16-row prime chunk now takes the `grid.y = m` dp4a program its 17-row sibling takes; decode and verify keep the tier (verify-exact has precedence; the exact-16 batched-decode tier and the K = 15 verify are untouched). The continuation gate no longer exempts a 16-row tail: a differing 16-row split fails it. The schedule-side alternative (fold the 16-row tail, refuse a 16-row restore) was rejected in the decision note: it leaves the engine's own continuation invariant broken at width 16.

**Gates, fixed tree, local RTX 5090 Laptop GPU (collector-locked, card alone before and after):**

- Two-scope walk: `WIDTH WALK scope=prime width 16 vs 17: 0 of 497 tensors differ` and `scope=prime width 48 vs 17: 0 of 497`; `scope=bare width 16 vs 17: 96 of 497` (the decode/verify class unchanged). Named site: `scope=prime layer 00 ssm_beta ... rows_differ=0/16 maxabs=0.000e0 ref_sha=e40f0aeaaec76762 sha=e40f0aeaaec76762 same`.
- This issue's table, seven arms, all `ok`, `A4 CONTINUATION GATE: PASS` on each: 9296 one call `14ab5f8b365dbd71`, `9280 + 16: logits_sha=14ab5f8b365dbd71 ok`, 48 to 208 `ok`; 9297 `e641952cac19e526` (17, 49 ok); 9311 `b61719f294866b05` (31, 63 ok); 9312 `736a88d7c7448dcb` (32, 64, 96 ok); `MEMRA_PRIME_CHUNK=32` one call `14ab5f8b365dbd71` (was `35bd15f063bfd5ba`: the cold prime is chunk-invariant at width 16 now); `MEMRA_PRIME_CHUNK=16` `bafe0e0a09a3d0a4` (both splits ok); `MEMRA_NO_BATCHED=1` identical to the default.
- `kernel-check`, both manifests: `ALL GREEN (109 cells, 10 skipped)`.
- `tools/spec-on-cache-hit-gate.sh qwen`: `SPEC-ON-CACHE-HIT GATE: ALL GREEN (qwen)` (the external drafter is absent on this rig; WARNING branch, as day 19).
- Twin gate: `PREFIX-NEWEST-TURN-FITS: ... cached_ok=7/7 lines_ok=8/8 ... identity_ok=8/8 grid_ok=21/21 grid=32 off_grid_calls=0 V1=ok V2=ok V3=ok V4=ok V5=ok V6=ok -> PASS`.
- Base tree (the fix's hunks reverted, nothing else) versus fix, same rig and artifact: default 9296 one call `14ab5f8b365dbd71` on both; 9297 `e641952cac19e526` on both; `MEMRA_PRIME_CHUNK=32` one call base `35bd15f063bfd5ba`, fix `14ab5f8b365dbd71` (the wide-chunk digest). The fix moved exactly the 16-row chunk program.

**Target card (one RTX PRO 6000 Blackwell, built on the box):** the same seven arms, every line digest-identical to the local table (`pro-single-day22/fix-gate/`). `kernel-check` there: 362 cells OK, 0 FAIL, one required manifest cell (`DUAL-BATCHED-AUX`) absent because the 9B artifact is not on the box, so the verdict line is owed (a step35-only rerun was refused six times on a lock held by another lane).

**Owed to the integ before main:** `run-gen` on a 16-token prompt and a t = 16 mod 4096 prompt, `run-spec` K=1..8, the target-card `kernel-check` line, `docs/TESTING.md` listing the 16-row split as a verdict. Not armed, stated: `prime_layers_gemma` and `step35_prime_cache_batch` do not take the scope; a quantized `out_f < 128` projection on those walks would still have the tier's program at 16 rows, and arming them is its own change with its own gate.

**Comment by avifenesh** [comment]
Lane B day 23 (`lane/spill-b-20260919`): the #427 battery re-run on main's mechanism.

#614 (`653c997f4`) fixes this at the tier: `Engine::small_m_tier_max()` is 16 inside the verify-exact scope and `PRIME_MIN_T - 1` outside it, so a 16-row prime call rides the same program as every wider prime for every caller of `matmul`/`matmul_pre`. The lane's day-22 `prefill_rows` scope was a second mechanism for the same behaviour and is dropped (lead ruling 26); the day-22 evidence stands (the width walk that named `ssm_beta`/`ssm_alpha` and the batched tier at m = 16, `research/spill-b-20260919/DAY22.md`). Decision record with both mechanisms and the two scope-gap sites read: `research/spill-b-20260919/PRIME-MIN-T-DECISION.md`, "Superseded by #614".

Receipts on the merged tree (main `653c997f4` + the lane's one-arm `qwen-a4-width-walk` diagnostic), `research/spill-b-20260919/DAY23.md`, every cell pass/fail through the collector, `executed-not-qualified`:

Local RTX 5090 Laptop GPU (`rtx5090-day23/`), served `Qwen3.8-27B-NVFP4-Q5K-mtp.gguf`:
- `WIDTH WALK width 16 vs 17: 0 of 497 tensors differ; tensor names: {}; sites: []` (no scope; 48 vs 17 also 0)
- seven-arm continuation table: every split `ok` at the one-call digest, `A4 CONTINUATION GATE: PASS` on all seven arms; `9280 + 16: logits_sha=14ab5f8b365dbd71 ok`; `MEMRA_PRIME_CHUNK=32` one-call `14ab5f8b365dbd71` (the default schedule's digest, was `35bd15f063bfd5ba` before the fix); digest for digest the day-22 fix table
- `kernel-check` both manifests: `ALL GREEN (109 cells, 10 skipped)`
- #379 hit gate: `SPEC-ON-CACHE-HIT GATE: ALL GREEN (qwen)`
- twin gate: `PREFIX-NEWEST-TURN-FITS: ... identity_ok=8/8 grid_ok=21/21 ... V1=ok V2=ok V3=ok V4=ok V5=ok V6=ok -> PASS`
- `run-gen`: `MATCH` (and `batched-prime ... MATCH`) on the raw-id validation prompt, the 90-token probe, a 16-token prompt and a 4112-token prompt (16 mod 4096)
- `run-spec` K=1..8 with the mint's embedded NextN drafter: `=== SELF-CONSISTENCY PASS ===` on the probe, the 16-token and the 4112-token prompt, acceptance > 0 at every K

Target card, one RTX PRO 6000 Blackwell (`pro-single-day23/`):
- `kernel-check`, step35 manifest alone and both manifests: `ALL GREEN (109 cells, 6 skipped)`, `DUAL-BATCHED-AUX [NVFP4 rp] out=48 m=3: bit-bad=0/0 OK` (the 9B artifact the 27b manifest needs was staged from the verified local copy, SHA-256 `52c9cceb...` equal to five repo receipts; manifests and checker unedited)
- seven-arm table, `run-gen` and `run-spec`: the same verdicts, argmaxes and accepted/drafted counts as the local card, digest for digest

Nothing is red against #614. Still owed to a rig that holds a Step-3.7-Flash GGUF: the step35 continuation split (16 vs 17 rows), a confirmation under #614 since the bound covers `step35_prime_cache_batch` by construction. Left open for the lead to close with the integ.

**Comment by avifenesh** [comment]
Two-card gate receipts for the #614 mechanism, integrated as #616 (merge `e2e9e294a`), lane B days 22 and 23 (`research/spill-b-20260919/DAY22.md`, `DAY23.md`, `PRIME-MIN-T-DECISION.md` section "Superseded by #614"): kernel named as the b16 batched-MMVQ tier at m = 16 on the `out_f < 128` GDN projections (`ssm_alpha`, `ssm_beta`); on the merged tree, local RTX 5090 and one RTX PRO 6000 Blackwell: width walk `16 vs 17: 0 of 497 tensors differ`, continuation gate seven arms every split `ok` (`9280 + 16: logits_sha=14ab5f8b365dbd71 ok`), kernel-check ALL GREEN both manifests (target card with the 9B staged, `DUAL-BATCHED-AUX ... OK`), hit gate ALL GREEN, twin gate PASS, run-gen MATCH and run-spec K=1..8 PASS on the standard, 16-token and 4112-token prompts. Owed as a confirmation, not a blocker: the step35 split cell (`step35_prime_cache_batch` passes m = 16 at B=1 with `attn_gate` below the GEMM floor; no Step-3.7-Flash GGUF on either rig today).
